### PR TITLE
[3101] Make it possible to hide cards in Deck representation

### LINKED ISSUE #3101: Add a Lane context menu to hide/reveal cards in Deck
* [ ] **I have checked that this feature has not yet been suggested by someone else.**

### Expected behavior

The idea is to make possible to hide/reveal some cards by using a context menu on a lane.
The context menu will be accessible through a "more" button on the right hand side of the lane header.
...

### Cuttings backs

...

### Potential side effects

...

### Describe alternatives you've considered

...

### PR DESCRIPTION
**Pull request by florianbarbin**
Bug: https://github.com/eclipse-sirius/sirius-web/issues/3101

# Pull request template

## General purpose
What is the main goal of this pull request?
- [ ] Bug fixes
- [x] New features
- [ ] Documentation
- [ ] Cleanup
- [ ] Tests
- [ ] Build / releng

## Project management
- [ ] Has the pull request been added to the relevant project and milestone? (Only if you know that your work is part of a specific iteration such as the current one)
- [ ] Have the `priority:` and `pr:` labels been added to the pull request? (In case of doubt, start with the labels `priority: low` and `pr: to review later`)
- [ ] Have the relevant issues been added to the pull request?
- [ ] Have the relevant labels been added to the issues? (`area:`, `difficulty:`, `type:`)
- [ ] Have the relevant issues been added to the same project and milestone as the pull request?
- [ ] Has the `CHANGELOG.adoc` been updated to reference the relevant issues?
- [ ] Have the relevant API breaks been described in the `CHANGELOG.adoc`? (Including changes in the GraphQL API)
- [ ] In case of a change with a visual impact, are there any screenshots in the `CHANGELOG.adoc`? For example in `doc/screenshots/2022.5.0-my-new-feature.png`

## Architectural decision records (ADR)
- [ ] Does the title of the commit contributing the ADR start with `[doc]`?
- [ ] Are the ADRs mentioned in the relevant section of the  `CHANGELOG.adoc`?

## Dependencies
- [ ] Are the new / upgraded dependencies mentioned in the relevant section of the `CHANGELOG.adoc`?
- [ ] Are the new dependencies justified in the `CHANGELOG.adoc`?

## Frontend

This section is not relevant if your contribution does not come with changes to the frontend.

### General purpose
- [ ] Is the code properly tested? (Plain old JavaScript tests for business code and tests based on React Testing Library for the components)

### Typing
We need to improve the typing of our code, as such, we require every contribution to come with proper TypeScript typing for both changes contributing new files and those modifying existing files.
Please ensure that the following statements are true for each file created or modified (this may require you to improve code outside of your contribution).

- [ ] Variables have a proper type
- [ ] Functions’ arguments have a proper type
- [ ] Functions’ return type are specified
- Hooks are properly typed:
	- [ ] `useMutation<DATA_TYPE, VARIABLE_TYPE>(…)`
	- [ ] `useQuery<DATA_TYPE, VARIABLE_TYPE>(…)`
	- [ ] `useSubscription<DATA_TYPE, VARIABLE_TYPE>(…)`
	- [ ] `useMachine<CONTEXT_TYPE, EVENTS_TYPE>(…)`
	- [ ] `useState<STATE_TYPE>(…)`
- [ ] All components have a proper typing for their props
- [ ] No useless optional chaining with `?.` (if the GraphQL API specifies that a field cannot be `null`, do not treat it has potentially `null` for example)
- [ ] Nullable values have a proper type (for example `let diagram: Diagram | null = null;`)

## Backend

This section is not relevant if your contribution does not come with changes to the backend.

### General purpose
- [ ] Are all the event handlers tested?
- [ ] Are the event processor tested?
- [ ] Is the business code (services) tested?
- [ ] Are diagram layout changes tested?

### Architecture
- [ ] Are data structure classes properly separated from behavioral classes?
- [ ] Are all the relevant fields final?
- [ ] Is any data structure mutable? If so, please write a comment indicating why
- [ ] Are behavioral classes either stateless or side effect free?

### How to test this PR?
_Please describe here the various use cases to test this pull request_

- [ ] Has the Kiwi TCMS test suite been updated with tests for this contribution?

### CODE REVIEW

File: packages/deck/backend/sirius-components-deck/src/main/java/org/eclipse/sirius/components/deck/renderer/component/CardComponent.java [hunk 1]
```diff
@@ -65,7 +68,26 @@ private Element doRender(VariableManager childVariableManager) {
...
+                .findFirst();
+        String cardId = optionalPreviousCard.map(Card::id).orElse(UUID.randomUUID().toString());
+        boolean visible = this.computeVisibility(optionalPreviousCard);
+
+        CardElementProps cardElementProps = new CardElementProps(cardId, cardDescription.id(), targetObjectId, targetObjectKind, targetObjectLabel, title, label, description, visible);
         return new Element(CardElementProps.TYPE, cardElementProps);
     }
+
+    private boolean computeVisibility(Optional<Card> optionalPreviousCard) {
+        if (optionalPreviousCard.isPresent()) {
```
Comment (frouene) [inline packages/deck/backend/sirius-components-deck/src/main/java/org/eclipse/sirius/components/deck/renderer/component/CardComponent.java]: An optional should not be used as a method argument, it's bad practice.
I prefer : 
```java
 boolean visible = optionalPreviousCard.map(this::computeVisibility).orElse(true);

...

 private boolean computeVisibility(Card previousCard) {
        return this.props.optionalDeckEvent().
                filter(ChangeCardsVisibilityDeckEvent.class::isInstance)
                .map(ChangeCardsVisibilityDeckEvent.class::cast)
                .map(ChangeCardsVisibilityDeckEvent::cardsVisibility)
                .map(cardsVisibility -> cardsVisibility.get(previousCard.id()))
                .orElse(previousCard.visible());
    }
```
What do you think ?

File: packages/deck/frontend/sirius-components-deck/src/card/DeckCard.tsx [hunk 2]
```diff
@@ -49,8 +48,8 @@ export const DeckCard = ({
   editable,
   t: translate,
 }: DeckCardProps) => {
-  const updateCard = (card: Card) => {
-    onChange(card);
+  const updateCard = (updatedCard: UpdateCardProps) => {
```
Comment (frouene) [inline packages/deck/frontend/sirius-components-deck/src/card/DeckCard.tsx]: Is it `update` or `updated` card ?

File: packages/deck/frontend/sirius-components-deck/src/laneHeader/useCardVisibility.ts [hunk 3]
```diff
@@ -0,0 +1,35 @@
...
+ * SPDX-License-Identifier: EPL-2.0
+ *
+ * Contributors:
+ *     Obeo - initial API and implementation
+ *******************************************************************************/
+
+import { useState } from 'react';
+import { Card } from '../Deck.types';
+import { UseCardVisibilityValue } from './useCardVisibility.types';
+export const useCardVisibility = (cards: Card[]): UseCardVisibilityValue => {
```
Comment (frouene) [inline packages/deck/frontend/sirius-components-deck/src/laneHeader/useCardVisibility.ts:18]: missing space after import

File: packages/deck/frontend/sirius-components-deck/src/representation/useDeckMutations.ts [hunk 4]
```diff
@@ -0,0 +1,370 @@
...
+  GQLEditCardData,
+  GQLEditCardVariables,
+  GQLEditDeckCardInput,
+  GQLEditDeckLaneInput,
+  GQLEditLaneData,
+  GQLEditLaneVariables,
+  GQLPayload,
+  UseDeckMutationsValue,
+} from './useDeckMutation.types';
+const deleteCardMutation = gql`
```
Comment (frouene) [inline packages/deck/frontend/sirius-components-deck/src/representation/useDeckMutations.ts:47]: missing space after import

**Review by frouene — CHANGES_REQUESTED** [review]

File: packages/deck/frontend/sirius-components-deck/src/card/DeckCard.tsx (hunk 2 above)
Comment (florianbarbin) [reply packages/deck/frontend/sirius-components-deck/src/card/DeckCard.tsx]: I will rename the argument ~cardToUpdate~ updateCardValue. This function will trigger the card updating

File: packages/deck/frontend/sirius-components-deck/src/laneHeader/useCardVisibility.ts (hunk 3 above)
Comment (florianbarbin) [reply packages/deck/frontend/sirius-components-deck/src/laneHeader/useCardVisibility.ts:18]: done

File: packages/deck/frontend/sirius-components-deck/src/representation/useDeckMutations.ts (hunk 4 above)
Comment (florianbarbin) [reply packages/deck/frontend/sirius-components-deck/src/representation/useDeckMutations.ts:47]: done

**Review by frouene — APPROVED** [review]

**Pull request merged**